### PR TITLE
Document the dependency graph

### PR DESCRIPTION
**Pull request by hevrard**
This change adds some documentation of the dependency graph: what it
is, and how it is created. It also adds and fixes some code comments
related to the dependency graph.

Bug: b/171775891

### CODE REVIEW

**Comment by hevrard** [comment]
You can see the README being rendered as a web page here:

https://github.com/hevrard/agi/blob/171775891-doc-depgraph/gapis/resolve/dependencygraph2/README.md

File: gapis/resolve/dependencygraph2/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+# Dependency graph
```
Comment (hevrard) [inline gapis/resolve/dependencygraph2/README.md:1]: It's probably easier to read here:

https://github.com/hevrard/agi/blob/171775891-doc-depgraph/gapis/resolve/dependencygraph2/README.md

File: gapis/resolve/dependencygraph2/README.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+> evolves, the few code snippets used here may not match the latest version,
+> but should still help to understand the high-level approach.
+
+In general, Command B depends on command A if command A has some side-effect to
+the API state or some resource (image, buffer...) memory that impacts command B.
+
+To take a simple GL example:
+
+```
+glClearColor(1, 0, 0, 1);     // Set clear color to red
```
Comment (yalcinmelihyasin) [inline gapis/resolve/dependencygraph2/README.md:15]: Should we give a Vulkan example instead?

**Review by yalcinmelihyasin — APPROVED** [review]
LGTM

File: gapis/resolve/dependencygraph2/README.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+## Creation of the dependency graph
+
+The dependency graph is created by mutating all the capture commands and
+monitoring their accesses to API state and memory.
+
+### Approach overview
+
+Let's give a high-level illustration of this process: consider we have a memory
+made of two slots. In each slot, we remember the ID of the node that was the
```
Comment (purvisa-at-google-com) [inline gapis/resolve/dependencygraph2/README.md:86]: Your first use of the word slot is to say "consider a memory made of two slots". Slot has not yet been defined.

File: gapis/resolve/dependencygraph2/README.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+In practice, alongside the `FragWatcher` and the `MemWatcher`, the dependency
+graph buidler also has a `ForwardWatcher` which handles forward dependencies via
+three callbacks: `OpenForwardDependency()`, `CloseForwardDependency()` and
+`DropForwardDependency()`.
+
+Most forward dependencies are only concerned with open and close operations. For
+instance, the mutation of `vkBeginCommandBuffer()` triggers
+`OpenForwardDependency()` which creates an open forward dependency, and
+`vkBeginCommandBuffer()` triggers `CloseForwardDependency()` which closes it,
```
Comment (purvisa-at-google-com) [inline gapis/resolve/dependencygraph2/README.md:154]: "`vkBeginCommandBuffer()` triggers `CloseForwardDependency()` which closes it,"

You mean:

"`vk**End**CommandBuffer()` triggers `CloseForwardDependency()` which closes it,"

?

File: gapis/resolve/dependencygraph2/memory.go [hunk 5]
```diff
@@ -99,17 +99,22 @@ func (b *memWatcher) OnReadObs(ctx context.Context, cmdCtx CmdContext, obs []api
...
+	// to avoid numerous realloc. This has a perceptible impact on big
+	// captures where it can save several seconds of computation.
+
 	// Compute the maximum possible of size of memAccesses at the end of `Flush`.
 	memAccessesCap := len(memAccesses)
 	for _, acc := range b.pendingAccesses {
 		memAccessesCap += len(*acc)
 	}
 
-	// Ensure that fragAccesses has sufficient capacity
```
Comment (purvisa-at-google-com) [inline gapis/resolve/dependencygraph2/memory.go:112]: good catch

File: gapis/resolve/dependencygraph2/memory.go [hunk 6]
```diff
@@ -197,6 +202,8 @@ func observationSlice(obs api.CmdObservation) memory.Slice {
 	return memory.NewSlice(obs.Range.Base, obs.Range.Base, obs.Range.Size, obs.Range.Size, obs.Pool, reflect.TypeOf(memory.Char(0)))
 }
 
+// applyMemWrite marks or creates the memoryWrites of wmap affected by (p, s)
```
Comment (purvisa-at-google-com) [inline gapis/resolve/dependencygraph2/memory.go:205]: wmap is writeMap? Feels worth documenting

File: gapis/resolve/dependencygraph2/memory.go [hunk 7]
```diff
@@ -218,6 +225,8 @@ func applyMemWrite(wmap map[memory.PoolID]*memoryWriteList,
 	return false
 }
 
+// applyMemRead returns the list of nodes for which there is a memoryWrite in
+// wmap that interesects with (p, s)
```
Comment (purvisa-at-google-com) [inline gapis/resolve/dependencygraph2/memory.go:229]: wmap as above

**Comment by purvisa-at-google-com** [comment]
LGTM with fixes

File: gapis/resolve/dependencygraph2/README.md (hunk 2 above)
Comment (hevrard) [reply gapis/resolve/dependencygraph2/README.md:15]: done

File: gapis/resolve/dependencygraph2/README.md (hunk 3 above)
Comment (hevrard) [reply gapis/resolve/dependencygraph2/README.md:86]: Clarified definition of slot in this context.

File: gapis/resolve/dependencygraph2/README.md (hunk 4 above)
Comment (hevrard) [reply gapis/resolve/dependencygraph2/README.md:154]: Good catch, thanks! corrected.

File: gapis/resolve/dependencygraph2/memory.go (hunk 6 above)
Comment (hevrard) [reply gapis/resolve/dependencygraph2/memory.go:205]: done

File: gapis/resolve/dependencygraph2/memory.go (hunk 7 above)
Comment (hevrard) [reply gapis/resolve/dependencygraph2/memory.go:229]: done

**Pull request merged**